File: Editorial/3578_Count_Partitions_With_Max_Min_Difference_at_Most_K/JavaScript/2_Monotonic_Queue_Optimization.js
var countPartitions = function (nums, k) {
    const n = nums.length;
    const mod = 1e9 + 7;
    const dp = new Array(n + 1).fill(0);
    const prefix = new Array(n + 1).fill(0);
    const minQ = [];
    const maxQ = [];

    dp[0] = 1;
    prefix[0] = 1;

    for (let i = 0, j = 0; i < n; i++) {
        // maintain the maximum value queue
        while (maxQ.length > 0 && nums[maxQ[maxQ.length - 1]] <= nums[i]) {
            maxQ.pop();
        }
        maxQ.push(i);
        // maintain the minimum value queue
        while (minQ.length > 0 && nums[minQ[minQ.length - 1]] >= nums[i]) {
            minQ.pop();
        }
        minQ.push(i);

        // adjust window
        while (
            maxQ.length > 0 &&
            minQ.length > 0 &&
            nums[maxQ[0]] - nums[minQ[0]] > k
        ) {
            if (maxQ[0] === j) {
                maxQ.shift();
            }
            if (minQ[0] === j) {
                minQ.shift();
            }
            j++;
        }

        const val = j > 0 ? prefix[j - 1] : 0;
        dp[i + 1] = (prefix[i] - val + mod) % mod;
        prefix[i + 1] = (prefix[i] + dp[i + 1]) % mod;
    }

    return dp[n];
};

File: Editorial/3501-3600/3578_Count_Partitions_With_Max_Min_Difference_at_Most_K/C/2_Monotonic_Queue_Optimization.c
#define MOD 1000000007LL

typedef struct {
    int* data;
    int front;
    int rear;
    int capacity;
} Deque;

Deque* createDeque(int capacity) {
    Deque* dq = (Deque*)malloc(sizeof(Deque));
    dq->data = (int*)malloc(capacity * sizeof(int));
    dq->front = 0;
    dq->rear = -1;
    dq->capacity = capacity;
    return dq;
}

int isEmpty(Deque* dq) { return dq->rear < dq->front; }

void pushBack(Deque* dq, int value) {
    if (dq->rear < dq->capacity - 1) {
        dq->data[++dq->rear] = value;
    }
}

void popBack(Deque* dq) {
    if (!isEmpty(dq)) {
        dq->rear--;
    }
}

void popFront(Deque* dq) {
    if (!isEmpty(dq)) {
        dq->front++;
    }
}

int front(Deque* dq) { return isEmpty(dq) ? -1 : dq->data[dq->front]; }

int back(Deque* dq) { return isEmpty(dq) ? -1 : dq->data[dq->rear]; }

int countPartitions(int* nums, int numsSize, int k) {
    long long* dp = (long long*)malloc((numsSize + 1) * sizeof(long long));
    long long* prefix = (long long*)malloc((numsSize + 1) * sizeof(long long));
    Deque* minQ = createDeque(numsSize);
    Deque* maxQ = createDeque(numsSize);

    dp[0] = 1;
    prefix[0] = 1;
    int j = 0;

    for (int i = 0; i < numsSize; i++) {
        // maintain the maximum value queue
        while (!isEmpty(maxQ) && nums[back(maxQ)] <= nums[i]) {
            popBack(maxQ);
        }
        pushBack(maxQ, i);

        // maintain the minimum value queue
        while (!isEmpty(minQ) && nums[back(minQ)] >= nums[i]) {
            popBack(minQ);
        }
        pushBack(minQ, i);

        // adjust window
        while (!isEmpty(maxQ) && !isEmpty(minQ) &&
               nums[front(maxQ)] - nums[front(minQ)] > k) {
            if (front(maxQ) == j) {
                popFront(maxQ);
            }
            if (front(minQ) == j) {
                popFront(minQ);
            }
            j++;
        }

        if (j > 0) {
            dp[i + 1] = (prefix[i] - prefix[j - 1] + MOD) % MOD;
        } else {
            dp[i + 1] = prefix[i] % MOD;
        }
        prefix[i + 1] = (prefix[i] + dp[i + 1]) % MOD;
    }

    int result = (int)dp[numsSize];

    free(dp);
    free(prefix);
    free(minQ->data);
    free(minQ);
    free(maxQ->data);
    free(maxQ);

    return result;
}
